lib/vkloader.c

```diff
@@ -146,38 +169,179 @@ ktxVulkanDeviceInfo_Construct(ktxVulkanDeviceInfo* This,
+ * function to provide it with the information.
+ *
+ * @param  This            pointer to the ktxVulkanDeviceInfo object to
+ *                        initialize.
+ * @param  physicalDevice handle of the Vulkan physical device.
+ * @param  device         handle of the Vulkan logical device.
+ * @param  queue          handle of the Vulkan queue.
+ * @param  cmdPool        handle of the Vulkan command pool.
+ * @param  pAllocator     pointer to the allocator to use for the image
+ *                        memory. If NULL, the default allocator will be used.
+ * @param pFunctions      pointer to the struct of functions used for
+ *
+ * @returns KTX_SUCCESS on success, KTX_OUT_OF_MEMORY if a command buffer could
+ *          not be allocated.
+ *
+ * @sa ktxVulkanDeviceInfo_destruct()
+ */
+KTX_error_code
+ktxVulkanDeviceInfo_ConstructEx(ktxVulkanDeviceInfo* This,
+                              VkInstance instance,
+                              VkPhysicalDevice physicalDevice, VkDevice device,
+                              VkQueue queue, VkCommandPool cmdPool,
+                              const VkAllocationCallbacks* pAllocator,
+                              const ktxVulkanFunctions* pFunctions)
+{
 
+#define LOAD_EXT_FUNC(member, fun) \
+do {                             \
+  if ((member).fun == NULL) {            \
+    (member).fun = (PFN_##fun)ktxLoadVulkanFunction(#fun); \
+    if ((member).fun == NULL) {          \
+        return KTX_FALSE;                               \
+    }\
+  } \
+} while (0)
+
+#define LOAD_INST_FUNC(member, instance, fun) \
+do {                             \
+  if ((member).fun == NULL) {            \
+    (member).fun = (PFN_##fun)((member).vkGetInstanceProcAddr)((instance), #fun); \
+    if ((member).fun == NULL) {          \
+        return KTX_FALSE;                               \
+    }\
+  } \
+} while (0)
+
+#define LOAD_DEVICE_FUNC(member, device, fun) \
+do {                             \
+  if ((member).fun == NULL) {            \
+    (member).fun = (PFN_##fun)((member).vkGetDeviceProcAddr)((device), #fun); \
+    if ((member).fun == NULL) {          \
+        return KTX_FALSE;                               \
+    }\
+  } \
+} while (0)
+
+
+    This->instance = instance;
     This->physicalDevice = physicalDevice;
     This->device = device;
     This->queue = queue;
     This->cmdPool = cmdPool;
     This->pAllocator = pAllocator;
-    vkGetPhysicalDeviceMemoryProperties(physicalDevice,
+
+    ktxVulkanFunctions funcs;
+    memset(&funcs, 0, sizeof(ktxVulkanFunctions));
+
+    if (pFunctions != NULL)
+        funcs = *pFunctions;
+
+    if (instance == VK_NULL_HANDLE || pFunctions == NULL) {
+        // This is the old behavior, where no functions where specified. We take a shot at loading them dynamically.
+#if defined(KTX_USE_FUNCPTRS_FOR_VULKAN)
+        // Delay loading not supported so must do it ourselves.
+        if (!ktxVulkanModuleHandle) {
```

Comment:
This if is not needed as `ktxVulkanLoadLibrary` returns success if the module handle is non-null. If you prefer the clarity of having it, please do the same in `ktxVulkanLoadFunction`.

Reply:
I'm not sure I understand this 😅 

I copied this from the old code, but you are saying that I can just delete it? The entire `#define` block?

Reply:
No. I am saying you can delete the `if (!ktxVulkanModuleHandle) {` and its closing `}`.

But now you mention it, there is no need for `KTX_USE_FUNCPTRS_FOR_VULKAN` any more. We now always use function pointers so `#if KTX_USE_FUNCPTRS_FOR_VULKAN` in any file, any `#else` clauses and the `#endif` can be deleted.